cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java

```diff
@@ -312,11 +311,24 @@ private HttpResponse doRequest(HttpMethod method, String resource,
     }
   }
 
-  private HttpRequest.Builder processBuilder(HttpRequest.Builder builder) {
-    if (SecurityRequestContext.getUserId() != null) {
-      builder.addHeader(Constants.Security.Headers.USER_ID, SecurityRequestContext.getUserId());
+  private HttpRequest.Builder addUserIdHeader(HttpRequest.Builder builder) throws IOException {
+    if (!securityEnabled) {
+      return builder;
+    }
+    // If the request originated from the router and was forwarded to any service other than dataset service, before
+    // going to dataset service via dataset service client, the userId could be set in the SecurityRequestContext.
```

Comment:
so whenever this client is used, the request originates at the router/app fabric. That's where it *could* be set in the ``SecurityRequestContext``. From here, it is forwarded to the dataset service, and during this operation, the USER_ID header *should* be set depending on what is received